### PR TITLE
Register Hibernate ORM user types for reflection

### LINKED ISSUE #2821: Hibernate ORM user types should be registered for reflection
From Zulip:
The attribute in my class is:
```
@Type(type = "jsonb")
private String payload;
```

I have declared the jsonb type:

```
@TypeDefs({
        @TypeDef(name = "jsonb", typeClass = RawJsonType.class)
})
```

```
Caused by: org.hibernate.MappingException: Could not determine type for: br.com.cresolconfederacao.idempotencia.hibernate.RawJsonType, at table: Registro, for columns: [org.hibernate.mapping.Column(payload)]
```

### PR DESCRIPTION
**Pull request by machi1990**
Fixes #2821.

### CODE REVIEW

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateUserTypeProcessor.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            final String typeClass = typeClassValue.asClass().name().toString();
+            userTypes.add(typeClass);
+        }
+
+        return userTypes;
+    }
+
+    private boolean isClass(String type) {
+        try {
+            Class.forName(type);
```
Comment (FroMage) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateUserTypeProcessor.java]: I think this should use jandex.

File: integration-tests/main/src/main/java/io/quarkus/it/jpa/CustomTypeEntity.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+
+import org.hibernate.annotations.Type;
+import org.hibernate.annotations.TypeDef;
+import org.hibernate.annotations.TypeDefs;
+
+@Entity
+@TypeDef(typeClass = BigIntType.class, name = "bigIntType1")
+@TypeDefs({
```
Comment (FroMage) [inline integration-tests/main/src/main/java/io/quarkus/it/jpa/CustomTypeEntity.java]: I think you should use three different user types, not three times the same one, to make sure you really register them all.

**Review by FroMage — DISMISSED** [review]
Thanks a lot, I have a few comments, but that's a very useful PR.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateUserTypeProcessor.java (hunk 1 above)
Comment (machi1990) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateUserTypeProcessor.java]: @FroMage Thanks. Do you think something like

```
 final String type = typeValue.asString();
 final DotName className = DotName.createSimple(type);
 if (index.getClassByName(className) == null) {
       continue; // Already registered through TypeDef annotation scanning
  }
```  
Will do the trick?

File: integration-tests/main/src/main/java/io/quarkus/it/jpa/CustomTypeEntity.java (hunk 2 above)
Comment (machi1990) [reply integration-tests/main/src/main/java/io/quarkus/it/jpa/CustomTypeEntity.java]: I can work on that.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateUserTypeProcessor.java (hunk 1 above)
Comment (FroMage) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateUserTypeProcessor.java]: I think it should.

File: integration-tests/main/src/main/java/io/quarkus/it/jpa/CustomTypeEntity.java (hunk 2 above)
Comment (machi1990) [reply integration-tests/main/src/main/java/io/quarkus/it/jpa/CustomTypeEntity.java]: Done.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateUserTypeProcessor.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        for (AnnotationInstance typeAnnotationInstance : typeAnnotationInstances) {
+            final AnnotationValue typeValue = typeAnnotationInstance.value(TYPE_VALUE);
+            if (typeValue == null) {
+                continue;
+            }
+
+            final String type = typeValue.asString();
+            final DotName className = DotName.createSimple(type);
+            if (index.getClassByName(className) == null) {
+                continue; // Already registered through TypeDef annotation scanning
```
Comment (gsmet) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateUserTypeProcessor.java]: I would change this comment to `Either already registered through TypeDef annotation scanning or not present in the index`

**Review by gsmet — CHANGES_REQUESTED** [review]
@machi1990 nice work as usual. I just added a very minor remark on a comment. Could you check?

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateUserTypeProcessor.java (hunk 3 above)
Comment (machi1990) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateUserTypeProcessor.java]: Done

**Review by gsmet — APPROVED** [review]

**Pull request merged**